packages/core/src/browser/pageState.spec.ts

```diff
@@ -0,0 +1,63 @@
+import { restorePageVisibility, setPageVisibility } from '../../test/specHelper'
+import type { PageExitEvent, PageState } from './pageState'
+import { createPageState } from './pageState'
+
+describe('createPageState', () => {
+  let pageState: PageState
+  let onExitSpy: jasmine.Spy<(event: PageExitEvent) => void>
+
+  beforeEach(() => {
+    pageState = createPageState()
+    onExitSpy = jasmine.createSpy()
+    pageState.onExit(onExitSpy)
+  })
+
+  afterEach(() => {
+    pageState.stop()
+    restorePageVisibility()
+  })
+
+  it('calls onExit listeners when the page is unloading', () => {
+    window.dispatchEvent(new Event('beforeunload'))
+    expect(onExitSpy).toHaveBeenCalledOnceWith({ isUnloading: true })
+  })
+
+  it('calls onExit listeners when the page becomes hidden', () => {
+    emulatePageVisibilityChange('hidden')
+    expect(onExitSpy).toHaveBeenCalledOnceWith({ isUnloading: false })
+  })
+
+  it('calls onExit listeners every time the page is unloading', () => {
+    window.dispatchEvent(new Event('beforeunload'))
+    window.dispatchEvent(new Event('beforeunload'))
+
+    expect(onExitSpy).toHaveBeenCalledTimes(2)
+  })
+
+  it('does not call onExit listeners when the visibility changes after the page is unloading', () => {
+    window.dispatchEvent(new Event('beforeunload'))
+    emulatePageVisibilityChange('hidden')
+    emulatePageVisibilityChange('visible')
+    emulatePageVisibilityChange('hidden')
+
+    expect(onExitSpy).toHaveBeenCalledOnceWith({ isUnloading: true })
+  })
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏any reason to stop calling onExit listeners after a first before unload?
FMU, after a canceled before unload, a user could stay on the page and continue to interact with it as it never tried to exit it.